### PR TITLE
feat(multi_object_tracker): add object class filtering in tracking process

### PR DESCRIPTION
**Pull request by technolojin**
## Description

The current object tracker uses the incoming classification directly.
When the detection has ambiguous classification, the classification result may changes in a short time.
Frequent classification change may have negative impact on the tracked object consumers (planner). 

In this PR, a simple classification filter is introduced.

  // 0. Normalize the input classification
  // 1-1. Update the matched classification probability with a gain (ratio of 0.05)
  // 1-2. If the label is not found, add it to the classification list
  // 2. Remove the class with probability < remove_threshold (0.001)
  // 3. Normalize tracking classification

## Related links

[TIER IV INTERNAL](https://tier4.atlassian.net/browse/RT1-5725)

## Tests performed

https://github.com/autowarefoundation/autoware.universe/assets/42434141/43ec72ca-951a-4585-a7c5-51f7ee0e06ec

Blue: Detected Object (input)
Yellow: Tracked Object (output)

The result shows the highest probable label is change only when the detection label is changed for a while.

## Notes for reviewers
This PR do not improve performance of detection classification.

## Interface changes
Not Applicable

## Effects on system behavior
Not Applicable

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: perception/multi_object_tracker/src/tracker/model/tracker_base.cpp [hunk 1]
```diff
@@ -54,6 +54,60 @@ bool Tracker::updateWithoutMeasurement()
...
+    }
+    // If the label is not found, add it to the classification list
+    if (!found) {
+      auto adding_class = new_class;
+      adding_class.probability *= gain;
+      classification_.push_back(adding_class);
+    }
+  }
+
+  // Normalize
```
Comment (YoshiRi) [inline perception/multi_object_tracker/src/tracker/model/tracker_base.cpp:101]: [imo]
Class normalization may used in input, so I think we can make this part to be a function. (lambda is OK)

File: perception/multi_object_tracker/src/tracker/model/tracker_base.cpp [hunk 2]
```diff
@@ -54,6 +54,60 @@ bool Tracker::updateWithoutMeasurement()
...
+    std::remove_if(
+      classification_.begin(), classification_.end(),
+      [](const auto & class_) { return class_.probability < 0.005; }),
+    classification_.end());
+
+  // Decay all existing probability
+  for (auto & class_ : classification_) {
+    class_.probability *= decay;
+  }
+
```
Comment (YoshiRi) [inline perception/multi_object_tracker/src/tracker/model/tracker_base.cpp:82]: [imo]
In my sense, input classification should be normalized too. How do you think about?

File: perception/multi_object_tracker/src/tracker/model/tracker_base.cpp [hunk 3]
```diff
@@ -54,6 +54,60 @@ bool Tracker::updateWithoutMeasurement()
...
+  // 1. Decay all existing probability
+  // 2. Update the matched classification probability with a gain
+  // 3. If the label is not found, add it to the classification list
+  // 4. Normalize
+
+  // Gain and decay
+  const double gain = 0.05;
+  const double decay = 1.0 - gain;
+
+  // If the probability is less than 0.005, remove the class
```
Comment (YoshiRi) [inline perception/multi_object_tracker/src/tracker/model/tracker_base.cpp:71]: How about put this after decay? or in the last part after normalization like pruning process. (Then we need normalization twice.)

**Review by YoshiRi — CHANGES_REQUESTED** [review]
Cool! Could you tell me your opinion about my questions?

File: perception/multi_object_tracker/src/tracker/model/tracker_base.cpp [hunk 4]
```diff
@@ -54,6 +54,60 @@ bool Tracker::updateWithoutMeasurement()
...
+  classification_.erase(
+    std::remove_if(
+      classification_.begin(), classification_.end(),
+      [](const auto & class_) { return class_.probability < 0.005; }),
+    classification_.end());
+
+  // Decay all existing probability
+  for (auto & class_ : classification_) {
+    class_.probability *= decay;
+  }
```
Comment (yukkysaito) [inline perception/multi_object_tracker/src/tracker/model/tracker_base.cpp:81]: Does it need to decay in time?
It is a little strange that the car becomes UNKNOWN when no observation.

Comment (yukkysaito) [reply perception/multi_object_tracker/src/tracker/model/tracker_base.cpp]: I understood

**Review by YoshiRi — APPROVED** [review]
LGTM. It worked within my environment.

**Pull request merged**